java-checks/src/main/java/org/sonar/java/checks/CompareWithEqualsVisitor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.sonar.java.model.declaration.MethodTreeImpl;
+import org.sonar.java.resolve.JavaType;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+
+public class CompareWithEqualsVisitor extends BaseTreeVisitor implements JavaFileScanner {
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    if (context.getSemanticModel() != null) {
+      scan(context.getTree());
+    }
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    if (!isEquals(tree)) {
+      super.visitMethod(tree);
+    }
+  }
+
```

Comment:
I would also override `visitBooleanExpression`, make it final to guarantee no change in subclasses, and only call an abstract method when its `==`or `!=`:
```java
  @Override
  public void visitBinaryExpression(BinaryExpressionTree tree) {
    super.visitBinaryExpression(tree);
    if (tree.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
      checkEquality(tree);
    }
  }

  public abstract void checkEqualityExpression(BinaryExpressionTree tree);
```

You can then simply implement `checkEqualityExpression` in each of the sub classes.